src/core/lib/event_engine/workqueue.h

```diff
@@ -0,0 +1,126 @@
+// Copyright 2022 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_CORE_LIB_EVENT_ENGINE_WORKQUEUE_H
+#define GRPC_CORE_LIB_EVENT_ENGINE_WORKQUEUE_H
+
+#include <grpc/support/port_platform.h>
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <atomic>
+#include <deque>
+
+#include "absl/base/thread_annotations.h"
+#include "absl/functional/any_invocable.h"
+#include "absl/types/optional.h"
+
+#include <grpc/event_engine/event_engine.h>
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/gprpp/time.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+// A fast work queue based lightly on an internal Google implementation.
+//
+// This uses atomics to access the most recent element in the queue, making it
+// fast for LIFO operations. Accessing the oldest (next) element requires taking
+// a mutex lock.
+class WorkQueue {
+ public:
+  // comparable to Timestamp::milliseconds_after_process_epoch()
+  static const int64_t kInvalidTimestamp = -1;
+
+  class Storage {
+   public:
+    Storage() = default;
+    // Take a non-owned Closure*
+    // Requires an exec_ctx on the stack
+    // TODO(ctiller): replace with an alternative time source
+    explicit Storage(EventEngine::Closure* closure) noexcept;
+    // Wrap an AnyInvocable into a Closure.
+    // The closure must be executed or explicitly deleted to prevent memory
+    // leaks. Requires an exec_ctx on the stack
+    // TODO(ctiller): replace with an alternative time source
+    explicit Storage(absl::AnyInvocable<void()> callback) noexcept;
+    ~Storage() = default;
+    // not copyable
+    Storage(const Storage&) = delete;
+    Storage& operator=(const Storage&) = delete;
+    // moveable
+    Storage(Storage&& other) noexcept;
+    Storage& operator=(Storage&& other) noexcept;
+    // Is this enqueued?
+    int64_t enqueued() const { return enqueued_; }
+    // Get the stored closure, or wrapped AnyInvocable
+    EventEngine::Closure* closure();
+
+   private:
+    EventEngine::Closure* closure_ = nullptr;
+    int64_t enqueued_ = kInvalidTimestamp;
+  };
+
+  WorkQueue() = default;
+  // Returns whether the queue is empty
+  bool Empty() const;
+  // Returns the number of elements in the queue
+  // This method locks the queue, be mindful of performance when using it.
+  size_t Size();
```

Comment:
What's the use case?

Reply:
Potential future work, helping queue-owners tune behavior based on queue depth. Given the expense, it may be best to remove it. An efficient implementation may be useful for prioritizing certain queues during work stealing, or deciding when to spin up more worker threads. Should we remove it as YAGNI?

Reply:
up to you

Reply:
Also nuked, it's not terribly complicated to add back.